heron/scheduler-core/src/java/com/twitter/heron/scheduler/LaunchRunner.java

```diff
@@ -134,40 +131,36 @@ public Boolean call() {
     // Set topology def first since we determine whether a topology is running
     // by checking the existence of topology def
     // store the trimmed topology definition into the state manager
+    // TODO(rli): log-and-false anti-pattern is too nested on this path. will not refactor
     result = statemgr.setTopology(trimTopology(topology), topologyName);
     if (result == null || !result) {
-      LOG.severe("Failed to set topology definition");
-      return false;
+      throw new LauncherException("Failed to set topology definition", topologyName);
     }
 
     result = statemgr.setPackingPlan(createPackingPlan(packedPlan), topologyName);
     if (result == null || !result) {
-      LOG.severe("Failed to set packing plan");
       statemgr.deleteTopology(topologyName);
-      return false;
+      throw new LauncherException("Failed to set packing plan", topologyName);
     }
 
     // store the execution state into the state manager
     ExecutionEnvironment.ExecutionState executionState = createExecutionState();
 
     result = statemgr.setExecutionState(executionState, topologyName);
     if (result == null || !result) {
-      LOG.severe("Failed to set execution state");
       statemgr.deletePackingPlan(topologyName);
       statemgr.deleteTopology(topologyName);
-      return false;
+      throw new LauncherException("Failed to set execution state", topologyName);
     }
 
     // launch the topology, clear the state if it fails
     if (!launcher.launch(packedPlan)) {
       statemgr.deleteExecutionState(topologyName);
       statemgr.deletePackingPlan(topologyName);
       statemgr.deleteTopology(topologyName);
-      LOG.log(Level.SEVERE, "Failed to launch topology");
-      return false;
+      throw new LauncherException("Failed to launch topology", topologyName);
     }
 
     return true;
```

Comment:
refactored so that `LaunchRunner` does not implement `Callable`. opened an issue related to https://github.com/twitter/heron/issues/1596.